lua/telescope/actions/utils.lua

```diff
@@ -0,0 +1,58 @@
+---@tag telescope.actions.utils
+
+---@brief [[
+--- Utilities to wrap functions around picker selections and entries.
+---
+--- Generally used from within other |telescope.actions|
+---@brief ]]
+
+local action_state = require('telescope.actions.state')
+
+local utils = {}
+
+--- Apply `f` to the entries of the current picker and return a table of mapped entries.
+--- - Note: mapped entries may include results not visible in the results popup.
+---@param prompt_bufnr number: The prompt bufnr
+---@param f function: Function to apply on entries of picker that takes (entry, index, row) as possible arguments
+---@return table: Result from `f` applied to entries
+function utils.map_entries(prompt_bufnr, f)
+  vim.validate{
+    f = {f, "function"}
+  }
+  local current_picker = action_state.get_current_picker(prompt_bufnr)
+  local results = {}
+  local index = 1
+  local row, result
+    -- indices are 1-indexed, rows are 0-indexed
+  for entry in current_picker.manager:iter() do
+    row = current_picker:get_row(index)
+    -- assign result to variable as function w/o return value
+    -- results in error upon table.insert as it's not technically nil but 'empty'
+    result = f(entry, index, row)
+    results[index] = result
+    index = index + 1
+  end
+  return results
+end
+
+--- Apply `f` to the multi selections of the current picker and return a table of mapped selections.
+---@param prompt_bufnr number: The prompt bufnr
+---@param f function: Function to apply on multi selection of picker
+---@return table: Result from `f` applied to selections
+function utils.map_selections(prompt_bufnr, f)
+  vim.validate{
+    f = {f, "function"}
+  }
+  local current_picker = action_state.get_current_picker(prompt_bufnr)
+  local results = {}
+  local result
+  for _, selection in ipairs(current_picker:get_multi_selection()) do
+    result = f(selection)
+    -- assign result to variable as function w/o return value
+    -- results in error upon table.insert as it's not technically nil but 'empty'
+    table.insert(results, result)
```

Comment:
Why not just `table.insert(results, f(selection))`

Reply:
Also the comment `assign result ...` does not apply here.

```lua
local results = {}

table.insert(results, nil)
table.insert(results, nil)
table.insert(results, 1)
table.insert(results, nil)
print(vim.inspect(results))
```

Reply:
This is the point, also applies to the remark below. Users might write `f` that don't return any value and this was meant to account for that in a simple manner.

```lua

-- doesn't work and raises, nothing to insert into table
x = {}
y = function() end
table.insert(x, y())
print(vim.inspect(x))

-- works
x = {}
y = function() return nil end
table.insert(x, y())
print(vim.inspect(x))

-- result is now nil
x = {}
y = function() end
result = y()
table.insert(x, result)
print(vim.inspect(x))
```

Reply:
Okay i see. Then just move the `local result` inside the for.

And maybe reframe the comment a little bit. Its was confusing for me :rofl:

Reply:
See discussion below, we should probably also track the index here and only insert results for key-value table for which the function returned non-`nil` values.